### PR TITLE
Initial MatMul implementation in IntrinsicArrayFunction

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**

### CODE REVIEW

**Comment by Thirumalai-Shaktivel** [comment]
TODO: Simplify the handling.

File: src/libasr/pass/intrinsic_array_function_registry.h [hunk 1]
```diff
@@ -1482,12 +1484,273 @@ namespace MinLoc {
...
+    static inline ASR::expr_t *instantiate_MatMul(Allocator &al,
+            const Location &loc, SymbolTable *scope,
+            Vec<ASR::ttype_t*> &arg_types, ASR::ttype_t *return_type,
+            Vec<ASR::call_arg_t> &m_args, int64_t overload_id,
+            ASR::expr_t *compile_time_value) {
+        if (compile_time_value) {
+            return compile_time_value;
+        }
+        /*
+         *   ------>
```
[the remark this answers is not in the document]
Comment (certik) [reply src/libasr/pass/intrinsic_array_function_registry.h:1624]:
```suggestion
         *   ------▶
```

**Comment by certik** [comment]
```fortran
function matmul(A, B) result(C)
real, intent(in) :: A(:,:), B(:,:)
real :: C(size(A,1), size(B,2))
end function
```

**Review by certik — APPROVED** [review]
I think this looks great.

File: src/libasr/pass/intrinsic_function.cpp [hunk 2]
```diff
@@ -299,6 +299,12 @@ class ReplaceFunctionCallReturningArray: public ASR::BaseExprReplacer<ReplaceFun
                     throw LCompilersException("Constant dimension cannot be extracted.");
                 }
             }
+        } else if ( dim_index == 2 ) {
+            result_var_ = PassUtils::create_var(result_counter,
+                std::string(ASRUtils::symbol_name(x->m_name)) + "_res",
+                x->base.base.loc, x->m_type, al, current_scope);
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_function.cpp:305]: This needs to be handled correctly, I will do this in the next PR.

File: tests/reference/asr-matrix_matmul_03-a99cdd1.stderr [hunk 3]
```diff
@@ -1,5 +1,5 @@
-semantic error: matmul accepts second matrix of type Logical if first matrix is of Logical type
- --> tests/errors/matrix_matmul_03.f90:5:24
+semantic error: The `matmul` intrinsic doesn't handle logical or complex type yet
+ --> tests/errors/matrix_matmul_03.f90:5:14
```
Comment (Thirumalai-Shaktivel) [inline tests/reference/asr-matrix_matmul_03-a99cdd1.stderr:2]: I have added this to the TODO

File: src/libasr/pass/intrinsic_array_function_registry.h [hunk 4]
```diff
@@ -1482,12 +1484,238 @@ namespace MinLoc {
 
 } // namespace MinLoc
 
+namespace MatMul {
+
+    static inline void verify_args(const ASR::IntrinsicArrayFunction_t &/*x*/,
+            diag::Diagnostics& /*diagnostics*/) {
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_array_function_registry.h:1490]: I will add some checks here, after that this is ready for review!

**Review by certik — APPROVED** [review]

**Comment by certik** [comment]
I think this looks good. I rebased it now to make sure everything works. If so, we can merge.

**Pull request merged**